Update: Add `always`/`never` option to `eol-last` (fixes #6938)

kdex opened the pull request:
In short, `eol-last`'s behavior could thus far only enforce that files ended with an OS-specific EOL. The rule's negation (i.e. that files should not end with an EOL) wasn't possible.

This PR adds a `"none"` option to `eol-last` which adds a way to check that a file doesn't end with an EOL, which fixes issue #6938. Additionally, there's also a fixer and some tests.

Here's a short overview of changes:
- Replaced ternary check of options with object access
- Initialized `location` with both `column` and `line` directly (note that `line` is now set by `sourceCode.lines.length` instead of performing a `.split("\n")`)
- Divided `checkBadEOF`'s code flow into two paths, where one path treats different operating systems' EOLs just like before, while the other path treats enforcing their absence
- Added a fixer that removes EOLs from the end of the file
- Added some tests for `"none"` option
- Extended `eol-last`'s documentation to mention `"none"`

kdex commented:
(The CLA error shouldn't be occurring; I've just signed it. Can I re-toggle it? Force-pushing doesn't help.)

platinumazure commented:
The CLA check doesn't like single names (kdex in your case). Some sort of "manual approval" on our end is theoretically possible (though I don't have access to do that), but your best bet is to change your git config to have something akin to a first and last name configured, amend and force-push the commit, and re-sign the CLA with the new name. Sorry for the inconvenience!

kdex commented:
@platinumazure Ah, thanks for the explanation! I'd really like to keep a global configuration for GitHub and work-related/private git projects; having multiple configurations set up would add unnecessary complexity to my backups (and break at other places, if I changed it globally). Is there someone I could ping for manual approval, possibly? I apologize!

platinumazure commented:
@eslint/eslint-tsc I'm not really sure what our CLA policy is really supposed to be here. The goal, as I understand it, is to ensure all code contributions, represented by commits, are owned by the contributors and freely licensed to us. We use name and email matching to trace "who provided the contribution" and match against the CLA list. Is "having two or more names" really an important part of this check? Do we, perhaps, expect a real name (which isn't the same as two or more names) in case we need to re-verify that the contribution was licensed according to our terms (or do other follow-up)? The jQuery Foundation has its reasons for doing these things- what about us, do we have to follow their norms 100% since we're now part of the Foundation?

pmcelhaney commented:
@kdex Can you not keep your global config and change it just for this project?

http://www.codeography.com/2011/08/05/project-specific-git-author.html

Otherwise, you should be able to override the author for just this commit.

```sh
git commit --amend --author="Your Name <your@email.com>"
```

pmcelhaney commented on lib/rules/eol-last.js:
```diff
@@ -34,23 +34,49 @@ module.exports = {
+                                start = match.index,
+                                end = sourceCode.text.length;
+
+                            return fixer.replaceTextRange([start, end], "");
```
Do we have tests for the fixer?

kdex commented:
@pmcelhaney If I were to change the git configuration just for this project, I'd have to back it up, or else I'd be facing the exact same problem in case of future PRs (or even any PR that I'm creating from another machine, really). That's what I was trying to get at when I was referring to it as "unnecessary complexity" on my side.

Again, I do apologize for making this more difficult than it needs to be, but I don't understand the motivation for matching git usernames against a certain RegExp, given that they doesn't even need to match real names.

About the fixer: Please correct me if I'm wrong, but I assumed that the fixer output corresponds to the `output` property in the `invalid` array in `tests/lib/rules/eol-last.js`, so I would think that the answer is "yes".

pmcelhaney commented:
I guess I don't understand what the unnecessary complexity is. If you sent the author configuration, what problem are you expecting to face next time?

(Nevermind the comment about tests. I figured it out about 10 seconds after I sent that comment.)

kdex commented:
@pmcelhaney Here's what I'd expect to happen:

-  If I were to make a PR on another machine, I'd just get another CLA error, since my git configuration would only have been configured locally
- Now I'm in a situation where I'd have to setup n different machines. If I don't remember the name that I used to match @jquerybot's CLA RegExp, I'd have to `$ git log | grep kdex` on that machine just to find out about the name I crafted for the CLA so I can re-configure git. Rinse and repeat for every time I format my hard drive.
- If, instead of grepping, I were to craft myself a new name every time and re-sign the CLA, that'd make the commit tree weird and inconsistent

Note that the CLA check can't check for real names anyway, making the necessity for the name to contain a space incomprehensible to me. Dropping this rule, everyone could commit from anywhere without the need for project-specific setups.

platinumazure commented:
@kdex I agree with you in spirit, if the only purpose of the CLA is to make sure that the same user who submitted commits has signed the CLA. That's why I've raised this to our TSC team. Please let's let them weigh this carefully and come to an informed decision.

gyandeeps commented:
@kdex Side note, I also manage 2 repos on my machine (work and public).

I have this bash function which get executed on startup which sets the parameters correctly for my work repos:

```sh

# inside .bashrc file
updateGitInfo() {
    originUrl=$(git ls-remote --get-url origin)
    #re='work github url match regex'
    
    if [[ -n $originUrl && $originUrl =~ $re ]]
    then
        echo Work git
        #git config user.email "your_email@example.com"
        #git config user.name "Awesome dude"
    fi    
}

updateGitInfo
```

My global info always has public github information.

vitorbal commented:
I think the CLA issue discussed here should have been fixed by the jQuery folks already. @platinumazure do you know if there's any way to force the bot to re-check the CLA?

platinumazure commented:
Just force-push, I think.

On Sep 1, 2016 3:03 PM, "Vitor Balocco" <notifications@github.com> wrote:

> I think the CLA issue discussed here should have been fixed by the jQuery
> folks already. @platinumazure <https://github.com/platinumazure> do you
> know if there's any way to force the bot to re-check the CLA?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/eslint/pull/6952#issuecomment-244195282>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AARWevPQoVeqNGAX7ChRtz6rkZsG3uf_ks5qly-VgaJpZM4JpOSZ>
> .
>

kdex commented:
Huh, seems like it still doesn't pass the CLA check. Are you sure that the change is live? [The signature page](https://contribute.jquery.org/CLA/) still mentions:

> The **name and email address** used to sign the CLA must match the name and email address in your git config.

I've also tried to re-sign the CLA, since I wasn't sure whether the back end just silently rejected all signatures without a space in the name, but to no avail.

vitorbal commented:
@dmethvin @kborchers hey all, sorry to bother you with this, the CLA check for @kdex still says "Manual verification required", even after forcing a re-check. Shouldn't the license check allow users with a single name now after https://github.com/jquery/jquery-license/pull/70?

kborchers commented:
My guess would be that the update didn't deploy due to the previous CLA disk issue. We will look into it ASAP.

kborchers commented:
OK, so it looks like the changes made in jquery/jquery-license#70 did not remove the 2 name check. That is done elsewhere. We'll have to get that change implemented but in the meantime I did the manual verification for @kdex and this is all good now.

kborchers commented:
I'll apologize for the noise in advance but we just pushed the real fix to the issue so I am going to remove the manual verification and recheck this PR. If it fails the CLA check I will re-validate the CLA signature and approve so you can merge and will test further elsewhere.

kborchers commented:
I think we're good. Sorry again for the noise and thanks!

platinumazure commented:
Thanks again for your help, @kborchers!

vitorbal commented:
Thanks a lot, @kborchers! :heart:

kborchers commented:
No problem :blush:

vitorbal commented on tests/lib/rules/eol-last.js:
```diff
@@ -54,6 +54,24 @@ ruleTester.run("eol-last", rule, {
+            code: "var a = 123;\r\n\r\n",
+            options: ["none"],
+            errors: [{ message: "Newline not allowed at end of file.", type: "Program" }],
+            output: "var a = 123;"
```
Could you add some tests for when there are other newlines in the code that are not EOF newlines?

vitorbal commented on tests/lib/rules/eol-last.js:
```diff
@@ -54,6 +54,24 @@ ruleTester.run("eol-last", rule, {
             options: ["windows"],
             errors: [{ message: "Newline required at end of file but not found.", type: "Program" }],
             output: "var a = 123;\r\n   \r\n"
+        },
```
Thanks for adding these! Could you add a couple of new tests for the "valid" scenarios as well?

vitorbal commented on docs/rules/eol-last.md:
```diff
@@ -8,7 +8,8 @@ as output files to the terminal without interfering with shell prompts.
 ## Rule Details
 
-This rule requires at least one newline at the end of non-empty files.
+This rule enforces at least one newline (or absence thereof) at the end
```
Could you update the title as well? "Require or disallow file to end with single newline"

vitorbal commented on lib/rules/eol-last.js:
```diff
@@ -20,7 +20,7 @@ module.exports = {
 
         schema: [
             {
-                enum: ["unix", "windows"]
```
Could you update the meta.docs.description to reflect the new option?

vitorbal commented on docs/rules/eol-last.md:
```diff
@@ -42,3 +42,4 @@ This rule has a string option:
 
 * `"unix"` (default) enforces line feed (LF) as newline
 * `"windows"` enforces carriage return line feed (CRLF) as newline
+* `"none"` enforces that the last line ends with neither LF nor CRLF as newline
```
I think this might read better for non-natives if it says "enforces that the last line does not end with a LF or CRLF as newline". In fact, I'm thinking maybe we could get rid of the "LF or CLRF" part altogether and just say: "enforces that the last line does not end with a newline". What do you think?

vitorbal commented:
@kdex thanks for bearing with us! I think this is already very good, just had a few nitpicks, a couple more updates for documentation and a suggestion for a few more tests. Thanks for your work so far!

kdex commented:
@vitorbal Thanks a lot for reviewing! I've updated the files accordingly and added five additional tests. Let me know if there's anything else that should be changed.

nzakas commented:
Should "none" be "never" to match our other rules?

platinumazure commented:
@nzakas I had a discussion with @kaicataldo about that. My reasoning (I favored "none") was that in this case, the single option described a newline style and needed to be an adjective, so "none" made more sense; whereas other cases are describing enforcement rate, and so needed to be an adverb (always/never).

Short version: Unix/windows/none felt like a more natural set to me than unix/windows/never.

I'm open to being persuaded otherwise, but at least @kaicataldo had agreed with me at one point.

vitorbal commented:
@kdex thanks! LGTM from my part.

kaicataldo commented:
I like "none" because, in the greater context of our rules, the presence of a "never" option seems to imply an "always" option to me (which was part of the discussion @platinumazure and I had), which this rule doesn't have.

Revisiting [this comment](https://github.com/eslint/eslint/issues/6938#issuecomment-242821557), I actually think if we want the greatest clarity and greatest consistency with our other rules we would want the options to be:
```js
    "no-multiple-empty-lines": ["error", "always"] // defaulting to "{ "style: "unix" }
    "no-multiple-empty-lines": ["error", "always", { "style": "windows" }]
    "no-multiple-empty-lines": ["error", "never"]
```

This could be done in a backwards compatible way.

kdex commented:
I agree with @kaicataldo. It'd be nicer if we could use this opportunity to make `eol-last` behave like most other rules.

Also, I noticed that `eol-last`'s currently shipped implementation is faulty when it comes to CRLF. More specifically, [these four tests](https://github.com/eslint/eslint/blob/ca3d448802d4e8a31d1a881226c0dcb65bf24971/tests/lib/rules/eol-last.js#L29-L32) are wrong, since they treat a Windows EOL just like a Unix EOL; all of these cases should therefore be invalid, as per default, Unix EOLs (LF) are expected.

My last commit fixes this behavior and therefore [adds two new cases](https://github.com/kdex/eslint/commit/c3500b038b1a574a818c6077250cdf5f7c7695c5#diff-ee4595dededf876c50a08c5e1f1a9ca7R92): The case where a CRLF was expected, but an LF was seen, and the case where an LF was expected, but a CRLF was seen. These new cases, again, come with tests and fixers.

Moreover, I also implemented @kaicataldo's suggestions in a backwards-compatible way, adding lots of [new tests](https://github.com/kdex/eslint/commit/c3500b038b1a574a818c6077250cdf5f7c7695c5#diff-9d43dd4092b307d22db09ba20b3645dd) that test both the old and new way of specifying options. Hence, I've removed the support for `"none"` and migrated the code over to `"always"`/`"never"`.

kdex commented:
I hope I'm not overhastily jumping to conclusions here. To me, it seemed that the general consensus was that "never" implies the existence of "always", which should have been implemented to begin with.

The only argument against `"always"`/`"never"` I can think of would be that the tests have become unnecessarily repetitive due to backwards-compatibility, which can be resolved by
- deprecating the old schema, and eventually removing support for it at some point
- re-introducing the `"windows"`, `"unix"`, `"none"` schema, accepting the inconsistency
- just being okay with tests being thorough :smile: 

Please let me know what the best approach would be here.

kaicataldo commented:
@kdex We need to make sure the other team members are on board - I'm sorry if I didn't make that clear. Let's see what the others have to say, but thanks for all your amazing work on this regardless!

pmcelhaney commented:
You're trying way too hard to shoehorn the always / never theme here. The question is what character(s) should go at the end of a file. Possible answers are "\r\n" (UNIX), "\n" (Windows), or "" (none, nothing, bare, empty, something along those lines). 

The rule makes perfect sense as it stands now. There's no need to change the semantics and introduce an eventual breaking change.

mysticatea commented:
I like `"always"`/`"never"` approach. And I think the `style` additional option is extra. It's overlapped with `linebreak-style` rule, IMO.

pmcelhaney commented:
Okay, if style is covered by another rule, that makes sense.

kaicataldo commented:
I'm :+1: for @mysticatea's suggestion. The current iteration of the rule only uses the `unix`/`windows` option for auto-fixing. The check itself is only for [if the last character of the source code is a `\n`](https://github.com/eslint/eslint/blob/master/lib/rules/eol-last.js#L44), meaning that deprecating `unix`/`windows` in favor of `always`/`never` would only affect the auto-fix (which is already covered by the `linebreak-style` auto-fix). We can ensure backwards compatibility by keeping the `unix`/`windows` options and making them enable the `always` option.

To sum it up: could we just add `always`/`never` options (that require or disallow a `\n` at the end of the source code and then in the docs deprecate `unix`/`windows` and suggest using the new `always`/`never` options in conjunction with `linebreak-style`?

kaicataldo commented:
@kdex Thanks for your patience. It looks like we have support [for this suggestion](https://github.com/eslint/eslint/pull/6952#issuecomment-244768743). Does that make sense? I think we should leave the current fixing behavior if either of the old options are used to ensure complete backwards compatibility. If the new options are used, the user can rely on `linebreak-style` to do the autofixing. 

Please ping back if I can help clarify anything :) Again, appreciate your work on this!

kdex commented:
@kaicataldo The current implementation in the PR should almost be equivalent to what you suggested. One difference, though, would be that `eol-last` itself expects an object with a `style` property where the user would pass his desired EOL flavor.

If I understand you correctly, I should remove this object and just assume that the newline flavor is `\n`, which will be fixed when used in conjunction with `linebreak-style`. (And of course, I should then update the documentation accordingly.) Is this correct?

kaicataldo commented:
Correct, the schema should just be:
```js
schema: [
  {
    enum: ["always", "never", "unix", "windows"]
  }
]
```

`"always"` and `"newline"` will always check for and require/disallow a `\n` (Unix-style linebreak) and autofix by adding or removing a `\n` character. 

If `"unix"` is specified, it will do the same check and autofix as `"always"`.

If `"windows"` is specified, it will do the same check as `"always"`  for a `\n` character, but autofix with the Windows-style linebreak of `\r\n` (for backwards compatibility).

kdex commented:
Thanks a lot for the clarification, @kaicataldo! :+1:

I've adjusted the schema to your description and was able to remove a lot of code/tests, relying on the fact that `linebreak-style` will cover us here. I've also updated the documentation accordingly.

Please let me know if I need to change anything. :)

kaicataldo commented on tests/lib/rules/eol-last.js:
```diff
@@ -26,10 +27,22 @@ ruleTester.run("eol-last", rule, {
         "var a = 123;\n\n",
         "var a = 123;\n   \n",
 
-        "\r\n",
```
I think we should leave these tests - even though we we're not enforcing a Windows linebreak style any longer, we want to make sure that this rule is still compatible with these cases

kaicataldo commented on tests/lib/rules/eol-last.js:
```diff
@@ -26,10 +27,22 @@ ruleTester.run("eol-last", rule, {
-        "var a = 123;\r\n",
-        "var a = 123;\r\n\r\n",
-        "var a = 123;\r\n   \r\n"
+        { code: "", options: ["unix"] },
```
Can we group together the tests that are for the deprecated options with a comment labeling them as such in both the valid and invalid tests (maybe at the end of each grouping)? Will make it much easier for the next person who might be removing the options in a future major version.

kaicataldo commented on docs/rules/eol-last.md:
```diff
@@ -1,4 +1,4 @@
-# Require file to end with single newline (eol-last)
+# Require or disallow file to end with single newline (eol-last)
```
I believe we are making these titles start with a lowercase character now - can someone else on the team  confirm?

kaicataldo commented on lib/rules/eol-last.js:
```diff
@@ -1,62 +1,95 @@
     meta: {
         docs: {
-            description: "enforce at least one newline at the end of files",
+            description: "enforce at least one newline (or absence thereof) at the end of files",
```
Can we make this consistent with the rule description in the title of the docs file? Thoughts on "require or disallow newlines at the end of files"?

kaicataldo commented:
Just a few small comments

kaicataldo commented on lib/rules/eol-last.js:
```diff
@@ -1,62 +1,95 @@
- * @fileoverview Require file to end with single newline.
+ * @fileoverview Require or disallow file to end with single newline.
  * @author Nodeca Team <https://github.com/nodeca>
+ * @author kdex <https://github.com/kdex>
```
We reserve the `@author` field for the file creator - your contribution will be in the Git history though!

kdex replied:
Sorry! I'll be removing mine. Note that I have seen cases where additional authors were added [years](https://github.com/eslint/eslint/blame/4126f12f792231c844859e13a3fbd6b7ff78c4ee/lib/rules/no-plusplus.js) after the file has been created, up to [this year](https://github.com/eslint/eslint/blame/c67632259783c9630831bce5f272eb6160ee086c/lib/rules/no-duplicate-case.js), which I followed. Did these cases slip under the radar?

platinumazure replied:
@kdex Yes, probably. Maybe we'll sweep through those later. (Some cases are legitimate collaborations at time of file creation, though.) Thanks for your understanding!

kaicataldo replied:
No need to apologize! Looks like some cases where we missed it, and should probably clean them up at some point.

kaicataldo commented on tests/lib/rules/eol-last.js:
```diff
@@ -1,6 +1,7 @@
- * @fileoverview Check, that file is ended with newline, and there are no multiple empty lines at the end.
+ * @fileoverview Tests for eol-last rule.
  * @author Nodeca Team <https://github.com/nodeca>
+ * @author kdex <https://github.com/kdex>
```
Same as https://github.com/eslint/eslint/pull/6952/files#r78205009

kdex commented:
@kaicataldo The issues from above should be fixed. One last thing: Is `@fileoverview` supposed to be consistent with either of `module.exports.meta.docs.description` or the description in the title of `eol-last.md`?

kaicataldo commented:
@kdex I think it makes sense for them to be, yes, though those seem to be all over the place in the codebase. Would you mind making them consistent?

kaicataldo commented:
Other than one tiny comment, LGTM. Waiting for a few others to review before merging. Thanks for being so responsive and for doing great work on this, @kdex! :+1:

kdex commented:
@kaicataldo This commit should make them more consistent, but having a look into the codebase, they're bound to be inconsistent anyway, as `@fileoverview`s start with capital letters and sometimes end with periods, whereas `meta.description` and documentation titles don't.

Having these two match implies that `@fileoverview` comments should also start in lowercase letters, which currently not the case.

If there are no objections, I would make all `@fileoverview` comments consistent for the entire codebase in another PR, making them start with lowercase letters and removing the periods.

Anyway, this PR should be ready-to-merge now, please review. :smile:

kaicataldo commented:
LGTM! Just want to make sure we get some other eyes on this before merging. Thanks again for contributing, @kdex!

Having the `@fileoverview` comments be consistent across all the rules seems like a nice thing to me - if that's something you'd like to do and you want to make an issue we can see what the rest of the team thinks.

kaicataldo commented:
@nzakas @vitorbal @mysticatea @platinumazure Would love at least one more pair of eyes on this before I merge :)

platinumazure commented on lib/rules/eol-last.js:
```diff
@@ -1,62 +1,94 @@
-                    linebreakStyle = context.options[0] || "unix",
-                    linebreak = linebreakStyle === "unix" ? "\n" : "\r\n";
+                    location = {
+                        column: 1,
```
Should this be zero-based? I thought I saw this come up elsewhere that columns should be zero-based on report and will automatically be converted to one-based before output.

kaicataldo replied:
Good catch, and yeah, `line` is one-based and `column` is zero-based: http://eslint.org/docs/developer-guide/working-with-rules#contextreport

platinumazure commented:
Just one question about the report column, otherwise LGTM.

vitorbal commented on tests/lib/rules/eol-last.js:
```diff
@@ -44,6 +63,52 @@ ruleTester.run("eol-last", rule, {
+            options: ["never"],
+            errors: [{ message: "Newline not allowed at end of file.", type: "Program" }],
+            output: "var a = 123;"
+        },
```
Could we add a test for two unix-style newlines being fixed to no newlines?

vitorbal commented:
This looks really good, just suggested one additional test case. Great work, @kdex!

kdex commented:
Thanks for reviewing, everyone!

@vitorbal Test added! Thank you.

@platinumazure That's quite a good catch, since neither assuming the column to be `0` nor `1` really makes sense for this rule. Come to think about it, the last EOL is really supposed to be the last character in a file, so it should be reported for: last line, last column, and **not**: last line, first column.

Another option would be to report for column `0`, but then I'd be forced to report the warning for a line number that doesn't even exist in the file (last line + `1`). Not sure if all editors can integrate that properly.

Making this change requires me to fix three tests in `cli-engine.js` and `eslint.js`.

Can I get some feedback about the expected semantics before I commit?

kaicataldo commented:
Last line, last column makes sense to me, since that's always the location this rule checks!

kaicataldo commented:
Actually, I'd be okay with us landing this as is (since the current reported column is the pre-existing behavior and not something you added), since you've done so much good work on it and I know there's been a lot of back and forth. Would hate to hold this up any more. We can always open another bug fix issue for the location if it is indeed incorrect.

kdex commented:
Agreed! Here's the test @vitorbal mentioned.

I already fixed the column issue in my local version; will commit and open a new issue/PR for that once we get this one merged. :smile:

kaicataldo commented:
Having the reported location be the last character of the file makes sense to me - let's see what the rest of the team has to say :)

This looks really great! Thanks again @kdex for your patience with the back and forth and for being so diligent with this PR - I think we ended up finding a much better solution through the process. Thanks!

vitorbal commented:
Thanks for your contribution, @kdex! :tada: :tada:
